### DISCUSSION TITLE
Detecting LED vs. RGB LED

### DISCUSSION THREAD

**Discussion by dmpayton** [micropython/micropython#10503] [Hardware & Peripherals]
Hey MicroPython folks! I'm in a bit of a pickle of a jam, and could use some help.

I've been working on building IOT air quality monitors with a local environmental nonprofit. We've been using the [Wemos C3 Mini](https://www.wemos.cc/en/latest/c3/c3_mini.html) (v1.0.0) with the latest MicroPython 1.19.1, and we've distributed about 50 of 'em in Central California.

Wemos recently started shipping a new version of this board, v2.1.0, which is _fine_, except for the LED differences – the 1.0.0 boards shipped with a single blue LED, while the 2.1.0 boards have an RGB LED, and both boards have the LEDs on pin 7. There's not a lot that's changed in the schematics, other than the antenna and the LED. 

On the v1.0.0 boards, I was using the blue LED as a general system status indicator, pulsing and blinking for various tasks and errors. I'm really hoping to be able to use the RGB LED on the newer boards to also indicate air quality at-a-glance.

My code needs to support both versions of the board because OTA updates pull from the same git repo and it feels crazy to run separate builds just to control an LED.

I was hoping to find a way to differentiate between the boards – a version number or _something_, somewhere in `sys`, `os.uname`, `esp32`, etc – but have come up short. I was wondering if there's some way to, I dunno, inspect the pin to see how it behaves so I can determine the board version and use the right led driver at runtime?

I'm a fair bit stuck on this right now, so any tips, ideas, suggestions or general advice would be super appreciated! Thank you!

**Comment by robert-hh** [comment]
You could try to set Pin7 to INPUT mode and read back the input value. Maybe there is a difference between the two boards. but chances are low. So the only choice would be to put some kind of configuration file on the board, telling which led to use, or use two different small python scripts to lite up the LED.

**Comment by cve2022** [comment]
Is the board assembled over a baseboard? If affirmative, depending on what is connected to pins A0 to A5, you could try detecting the board type, as the order of these pins is different. Or you could add some pull-up/down resistor to one of the pins on the boards with the newest module and detecting the level by setting the pin's internal pull-down/up.

**Comment by sosi-deadeye** [comment]
You could use a simple DIP-Switch to set the hardware-configuration.

**Comment by jimmo** [comment]
@dmpayton  Wow, that's quite a frustrating change!

You would need to test this, but perhaps you could just assume that it's the 2.x hardware and drive the neopixel, but then leave the pin in the state suitable for the 1.x hardware. i.e. When you want to change the LED state, send the appropriate WS2812 signal (using the `neopixel` module, or `machine.bitstream` directly), then leave the pin high or low.

I'm pretty sure the WS2812 won't care about the state of the pin being left high or low.

**Reply by pdw-mb** [reply]
Neat idea!  I think you will probably need to set the output low for ~300us before sending a new WS2812 signal in order to avoid the long "1" being treated as the first bit of the next value.  

If you were planning to drive the simple LED with PWM rather than on/off, it gets a bit more interesting but I think you could make it work.

**Reply by dmpayton** [reply]
I really like this idea! I posted an update below with my (current) solution, but I'd love to explore this idea in the future. Unfortunately, I'm using PWM to fade and pulse the LED, so that should be an adventure. :)

**Comment by dmpayton** [comment]
Thank you all for the suggestions, there are definitely some things I want to try in a future hardware iteration!

@mattytrentini was kind enough to [respond to me on Twitter](https://twitter.com/matt_trentini/status/1614583277079363584), suggesting to set the pin to output high, switch to input, and time the voltage drop. Here's my implementation:

```python
def led_voltage_drop():
    pin = machine.Pin(7, machine.Pin.OUT)
    pin.on()
    pin.init(machine.Pin.IN)
    start = time.ticks_us()
    while pin.value() != 0:
        continue
    return time.ticks_diff(time.ticks_us(), start)
```

Open to any suggestions, but here are the results so far:

```
INFO:main:Wemos C3-Mini: 1.0.0
INFO:main:lead_voltage_drop(): 11587
INFO:main:lead_voltage_drop(): 13487
INFO:main:lead_voltage_drop(): 13731
```

```
INFO:main:Wemos C3-Mini: 2.1.0
INFO:main:lead_voltage_drop(): 2617
INFO:main:lead_voltage_drop(): 2726
INFO:main:lead_voltage_drop(): 2628
```

There's definitely an overall difference between the two, and the numbers appear consistent across a few boards of each version that I've tried. Huzzah!

It should now be just a simple check to get the version:

```python
def board_version():
    # 8,000 is roughly in the middle between the two LEDs
    if led_voltage_drop() < 8000:
        # Usually < 3,000
        return '2.1.0'
    # Usually > 10,000
    return '1.0.0'
```

Does this seem like a reasonable solution and implementation? I'm newer to EE (my background is mostly web dev), so I don't know if there are any other considerations that should be taken into account here. Is there anything else that might affect the voltage drop duration, causing my code to set the wrong version?

**Reply by cve2022** [reply]
@dmpayton , inventive, but IMO it may not be very robust on the long term or with different batches, because it relies on variables you may have no control over, such as ambient temperatur, humidity, variability of the components (LED and RP2 itself) between different batches, that can alter the leakage currents that cause the stray capacitance to discharge after your probing.

> There's not a lot that's changed in the schematics, other than the antenna and the LED.

Actually the A inputs are ordered differently.
![image](https://user-images.githubusercontent.com/108999978/212999974-4a2756cc-fb48-418c-bcba-342018c3d025.png) ![image](https://user-images.githubusercontent.com/108999978/213000091-c68eae1c-b4e9-4d91-a585-43313e1f4f70.png)

If you have some thing, say, a voltage divider, connected to one of them, that could solve your problem. 

How are you communicating with the sensor? Is the board assembled on some baseboard? If affirmative, you could add a resistor between one of the free IO pins and GND or 3V3 and use it to detect the V2.x board on the new monitors you ship.

Other than that, you may try following @jimmo 's and @pdw-mb 's suggestions: you may send the stream to the WS2812B  (it must take ~30us to send the 24 bits) then switch to PWM mode, starting the output high for as long as needed, then low for > 50us (see WS2812b data sheet) and repeat:
![image](https://user-images.githubusercontent.com/108999978/213008138-e2f0f362-4de5-417f-9d82-84d4bc0c914e.png)

Mind you, after receiving the 24 bits stream, the WS2812B outputs the remaining bits it receive to the next WS2812Bs in the chain, so its input staying in high level must not affect it. The low level time > 50us resets WS's state machine and then you need to send the new data again. I did not test it but I think it works, based on the (scarce) data sheet information. It is a little complicated for my taste but it may be worth trying.

You may also try what @jimmo suggested, after sending the stream to the WS2812B, you switch to PWM with much longer high and low times; maybe the LED will simply ignore the input. It is not strictly according to the protocol, so it may work now or if may stop working in future versions of the LED or for second sourced parts.

**Reply by cve2022** [reply]
> ...
> I'm pretty sure the WS2812 won't care about the state of the pin being left high or low.

On second thought, I think @jimmo is right: it is very likely that the machine state in the WS2812 will simply ignore if the pin is kept on a state longer than the timings of the data sheet: no level transition, no data shifting.

> You may also try what @jimmo suggested, after sending the stream to the WS2812B, you switch to PWM with much longer high and low times; maybe the LED will simply ignore the input.

Simpler and if it works, it solves your problem without resorting to a complicated scheme like the the one described above this  sentence and without having to make any change to the board(s). It is worth a shot.

**Reply by freemansoft** [reply]
I'm surprised they re-ordered the pins.  That seems like an unnecessary breaking change especially if there is no way to read the board version.

**Reply by pdw-mb** [reply]
> > You may also try what @jimmo suggested, after sending the stream to the WS2812B, you switch to PWM with much longer high and low times; maybe the LED will simply ignore the input.
> 
> Simpler and if it works, it solves your problem without resorting to a complicated scheme like the the one described above this sentence and without having to make any change to the board(s). It is worth a shot.

AIUI, the WS2812B latches and displays its current input whenever it sees a long low.  If you're feeding it a slow PWM, you're effectively sending it a single one bit, then latching.  I'm not sure exactly what it'll do, but I suspect that it will lose whatever it was supposed to be showing.

So I suspect that you will need to do what's shown in the diagram - slow PWM with a burst of WS2812 data at the start of each pulse, and make sure that you don't use a duty cycle that requires an off time of less than the WS2812B reset time.  Apparently some WS2812B variants have a longer reset time (300us rather than 50us)

**Reply by cve2022** [reply]
> AIUI, the WS2812B latches and displays its current input whenever it sees a long low. If you're feeding it a slow PWM, you're effectively sending it a single one bit, then latching. I'm not sure exactly what it'll do, but I suspect that it will lose whatever it was supposed to be showing.

I don't think so, based on the data sheet.

![image](https://user-images.githubusercontent.com/108999978/213145949-278aed01-3fb5-4aa6-ab47-d8b6d212e4f1.png)

Once the first WS in the chain gets its 24 bits, it latches the RGB word and forwards the incoming bits to its DO output for the next in the chain its capture its RGB bits and so forth.

Ones and zeros of each bit are determined by the width of the TxH pulse, IOW, the point at which the negative edge occurs.
![image](https://user-images.githubusercontent.com/108999978/213146564-732efc41-94d2-466e-a6e4-244e15d2b12c.png)

TH+TL must be 1.25us +/- 600ns per the data sheet. My guess is the WS ignores negative edges after long TxH pulses, but that needs to be tested. It is worth testing it, as it is way simpler than interpersing the RGB stream with the PWM pulses.

**Reply by pdw-mb** [reply]
[This page](https://wp.josh.com/2014/05/13/ws2812-neopixels-are-not-so-finicky-once-you-get-to-know-them/) says:

> There is a maximum time the level can stay low (TxL) before a reset is triggered and any loaded data is latched and displayed on the LED.

Which implies that the LED only updates when it sees the reset (meaning that all LEDs in the strip update simultaneously).  Further down the page the author appears to confirm this using a photo-detector.  

As you say, it needs some testing.  I'm quite tempted to give it a go :-)

**Reply by mattytrentini** [reply]
Just a quick note to be wary about strictly relying on behaviour described in the datasheet - there are so many WS2812B clones that all differ in _slight_ ways as soon as you stray from the norm!

**Reply by robert-hh** [reply]
I just made a test with a neopixel ring.
The WS2812 update when they get a proper pattern. They do not need a reset, and my test code does not apply a reset. And when the LED's are displaying, the reset pulse does not change anything. The WS2812 ignores any longer 1 or 0 level at the input, which does not look like a data frame.

**Reply by cve2022** [reply]
> I just made a test with a neopixel ring.

Would you mind sharing your code? I am curious to test overhere with pico and esp32. Would save me some time coding from scratch.

**Reply by cve2022** [reply]
> Just a quick note to be wary about strictly relying on behaviour described in the datasheet - there are so many WS2812B clones that all differ in _slight_ ways as soon as you stray from the norm!

I tend to agree, things may go awry when we use undocumented features that can change any time for an authentic part, even worse for clones.

**Reply by cve2022** [reply]
> I'm surprised they re-ordered the pins. That seems like an unnecessary breaking change especially if there is no way to read the board version.

My guess is they have some baseboard or shield for which the original pinout of the SPI (left) does not work and that was fixed with the new version (right).

![image](https://user-images.githubusercontent.com/108999978/213210850-a0b589b8-c6e6-4f91-991e-14353a028b0a.png) ![image](https://user-images.githubusercontent.com/108999978/213210681-8c5cfa6b-a668-4881-b9c8-2c8adea8e60d.png)

**Comment by robert-hh** [comment]
> Would you mind sharing your code?

See below. I used a  SAMD51 board. Pin D13 has a simple LED on the board. So the script fills the neopixels, blinks 3 times the LED, clears the neopixels, blinks the LED three times, ......
```
import time
import machine, neopixel

pattern = [
    (4, 0, 0),
    (20, 0, 0),
    (80, 0, 0),
    (200, 0, 0),
    (100, 0, 0),
    (30, 0, 0),
    (10, 0, 0),
    (1, 0, 0)]

def run(delay=500, neonum = 24):

    led = machine.Pin("D13", machine.Pin.OUT)
    np = neopixel.NeoPixel(led, neonum, timing = 1)
    for i in range(5):
        for j in range(neonum):
            np[j] = pattern[j % 8]
        np.write()
        for l in range(3):
            led(1)
            time.sleep_ms(delay)
            led(0)
            time.sleep_ms(delay)
        np.fill((0,0,0))
        np.write()
        for l in range(3):
            led(1)
            time.sleep_ms(delay)
            led(0)
            time.sleep_ms(delay)
```

**Reply by cve2022** [reply]
>>    Would you mind sharing your code?
>
>See below. I used a SAMD51 board. Pin D13 has a simple LED on the board. So the script fills the neopixels, blinks 3 times the LED, clears the neopixels, blinks the LED three times, ......

OK, I did some tests here, running the script on a Pico after modifying it. I noticed some visual anomaly. I modified the pattern to better visualize it (RGB combinations and much lower bright - the damn LEDs are too bright). I also added a version of the function that does not blink the led (`run2()`, not very creative I know) as a gold standard.

The aforementioned anomaly occurred with a 16-LEDs ring (there is a lag in the update of the first LED - the LEDs in it are most likely WS2812s  clones). If you call `run()` with higher delays, say, 2000ms, it is more noticeable. A different ring, with 24 LEDs works fine (though most likely also clones). 

The 16-LEDs ring works fine on an ESP32. Go figure.

```
import time
import machine, neopixel

BR = 5
pattern = [
    (BR, 0, 0),
    (0, BR, 0),
    (0, 0, BR),
    (BR, BR, 0),
    (BR, 0, BR),
    (0, BR, BR),
    (BR, BR, BR),
    (0, 0, 0),
    ]

def blank(np):
    np.fill((0,0,0))
    np.write()

def run(delay=500, neonum = 24):
    led = machine.Pin(0, machine.Pin.OUT)
    np = neopixel.NeoPixel(led, neonum, timing = 1)
    led_dly = int(delay/8)
    blank(np)
    time.sleep_ms(delay)
    for i in range(16):
        for j in range(neonum):
            np[j] = pattern[(i + j) % len(pattern)]
        np.write()
        for l in range(8):
            led(1)
            time.sleep_ms(led_dly)
            led(0)
            time.sleep_ms(led_dly)

def run2(delay=500, neonum = 24):
    led = machine.Pin(0, machine.Pin.OUT)
    np = neopixel.NeoPixel(led, neonum, timing = 1)
    blank(np)
    time.sleep_ms(delay)
    for i in range(16):
        for j in range(neonum):
            np[j] = pattern[(i + j) % len(pattern)]
        np.write()
        time.sleep_ms(delay)
        time.sleep_ms(delay)
```

**Reply by pdw-mb** [reply]
What's the nature of the anomaly?  Does the first pixel behave any differently to subsequent ones? 

I'd try adding a short delay (1ms should be plenty) after the `np.write()`

**Reply by cve2022** [reply]
> What's the nature of the anomaly? Does the first pixel behave any differently to subsequent ones?

There is a visual lag between the others updating and the first of the string. For example, if the sequence is R/G/B/etc. (like the pattern above), the next iteration should be G/B/Yellow/etc. all changing simultaneously, but what happens with that ring is that all but the first do update, so the second pattern is R, B/Yellow/etc. for a short time interval then the correct G/B/Yellow/etc. and so forth.

> 
> I'd try adding a short delay (1ms should be plenty) after the `np.write()`

I did it, it did fix the behavior.

I will do some scoping when I have time to see the timing differences between the ESP32 and the Pico to try to understand why this happens (out of sheer curiosity).

FWI, this is the final version of the test code:

```
import time
import machine, neopixel

BLANK_LEDS = 24
NLEDS = 24
BR = 5
pattern = [
    (BR, 0, 0),
    (0, BR, 0),
    (0, 0, BR),
    (BR, BR, 0),
    (BR, 0, BR),
    (0, BR, BR),
    (BR, BR, BR),
    (0, 0, 0),
    ]

def blank (np = None, neonum = BLANK_LEDS, fill = (0,0,0) ):
    if not np:
        led = machine.Pin(0, machine.Pin.OUT)
        np = neopixel.NeoPixel(led, neonum, timing = 1)
    np.fill(fill)
    np.write()
    time.sleep_ms(1)

def run(delay=500, neonum = NLEDS, led_blink = True):
    led = machine.Pin(0, machine.Pin.OUT)
    np = neopixel.NeoPixel(led, neonum, timing = 1)
    led_dly = int(delay/8)
    blank(np)
    time.sleep_ms(delay)
    for i in range(16):
        for j in range(neonum):
            np[j] = pattern[(i + j) % len(pattern)]
        np.write()
        time.sleep_ms(1)
        if led_blink:
            for l in range(8):
                led(1)
                time.sleep_ms(led_dly)
                led(0)
                time.sleep_ms(led_dly)
        else:
            time.sleep_ms(delay)
            time.sleep_ms(delay)

blank()
```

**Reply by pdw-mb** [reply]
I've just tried your code on a Pico, and get exactly the same results, which I think I can explain.

It sounds like the LEDs do only update once they get the reset/latch signal (low for > 5us - 300us, depending on LED flavour).  So the first LED only gets that after `led_dly` because there was no gap between the `np.write()` and setting the output to high again.  The first LED reshapes the pulse, so although it receives a `1` for `led_dly` (500ms), it'll pass on a pulse of the prescribed `1` width (800us), and then go low for the rest of the 500ms (and the next 500ms), and the subsequent LEDs will see a reset and update.  The first LED updates once it sees the `led(0)` 500ms later. 

If I add:

```py
led(1)
time.sleep(1)
```

immediately after the `np.write()`, the gap between subsequent and first LEDs updating grows - I can increase it to many seconds, and the LEDs stay out of sync for that number of seconds.

On my board, I need to add `time.sleep_us(60)` to get the LEDs in sync.  Would be interested to know what it takes on yours. 

According to the link I posted above, the latch/reset threshold can be as low as 5-6us.  My guess at the difference between the ESP32 and Pico is a slight difference in the time it takes to get from the last `0` output by `np.write()` to your code calling `led(1)`.  Try putting `led(1)` immediately after `np.write()` and see if that breaks the code on the ESP32.

**Reply by pdw-mb** [reply]
Here's some code you can use to find the minimum latching time on an LED:

```py
import time
import machine
import array

T_RESET_US=260

data = array.array('b', [0,0,1,0])
on = array.array('b', [10,10,10])

led = machine.Pin(11, machine.Pin.OUT)

while True:
    time.sleep(1)
    machine.bitstream(led, 0, (400, 850, 400, T_RESET_US * 1000), data)
    led(1)
    time.sleep(1)
    led(0)
    time.sleep(0.02)
    machine.bitstream(led, 0, (400, 850, 800, 450), on)
```

If `T_RESET_US` is less than the threshold, the LED won't latch the "off" pattern until `led(0)` is called, and will flash off briefly once every 2 seconds.  If `T_RESET_US` is more than the threshold, the LED should flash on for 1s, off for 1s.

I've tried this on a genuine (probably) WS2812B-2020 and the threshold is ~ 260us (datasheet says 280us, so that's good).

I tried it on some WS2812Bs of dubious origin, and the threshold is around 70us.

**Reply by cve2022** [reply]
I was guessing something in the line of what you described above as the cause of the issue.

> On my board, I need to add time.sleep_us(60) to get the LEDs in sync. Would be interested to know what it takes on yours.

With mine, it takes 70us for it to work properly in a consistent way (it fails sometimes with 60).

Adding led(1) to the ESP32 code does not cause the issue with one exception: only the very first pattern, after I stop the running code (with Thonny) and restart it again. The next patterns are OK without the lag.

**Reply by pdw-mb** [reply]
I'm not familiar with the ESP32, but I can see that the `bitstream` implementation is quite different and uses a feature called RMT, which I suspect cancels the pin mode of the GPIO pin, so the `led(1)` call isn't actually doing anything.  Try calling `machine.Pin(N, machine.Pin.OUT)` immediately after `np.write()` and before `led(1)`.

**Reply by cve2022** [reply]
> I'm not familiar with the ESP32, but I can see that the `bitstream` implementation is quite different and uses a feature called RMT, which I suspect cancels the pin mode of the GPIO pin, so the `led(1)` call isn't actually doing anything. Try calling `machine.Pin(N, machine.Pin.OUT)` immediately after `np.write()` and before `led(1)`.

FWIW whatever the driver does after the [RMT](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html) ends transmitting, the LED I added for the tests blinks.

**Reply by pdw-mb** [reply]
> FWIW whatever the driver does after the [RMT](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html) ends transmitting, the LED I added for the tests blinks.

Ah, I hadn't realised you had a test LED.  I guess it can't be that then.  Would be interesting to know what the output looks like on a scope.  Otherwise, it might be worth doing a crude timing check using `ticks_us()` to see if the `np.write()` is noticeably slower on the ESP32.

I had a play with this over the weekend on an RP2040 board, and got it working nicely with PWM, and had two LEDs fading in and out together on the same signal.  As you'd expect, max duty cycle is limited by the LED latch time, although for 70us latch time and 1kHz PWM, you can get up to 93% duty cycle, which is pretty usable.  It also needs a gap of the latch time inserting before the WS2812B data.

**Reply by cve2022** [reply]
> Ah, I hadn't realised you had a test LED. I guess it can't be that then. Would be interesting to know what the output looks like on a scope. 

Here you go. To avoid cluttering the post, I will post few images. The blue trace is from a GPIO I added to synchronize the scope. It is set right before the np.write() and reset right after it:
```
        sync(1)
        np.write()
        sync(0)
```

## ESP32
![image](https://user-images.githubusercontent.com/108999978/214577522-63f42571-452d-4c05-9f76-ca2103832841.png)

The delay between the sync GPIO set and the data stream start is 256us. The delay between the data stream end and the sync GPIO reset is 104us (notice that the yellow trace goes to high level right after it).

Here, one bit of the data stream (notice the 1.25us bit width):
![image](https://user-images.githubusercontent.com/108999978/214578417-60174b22-d601-4b09-8bca-303873469a16.png)

## RP2
![image](https://user-images.githubusercontent.com/108999978/214578939-a3e1c6c8-6010-4c15-a25a-a935e7eaf3d0.png)
![image](https://user-images.githubusercontent.com/108999978/214579281-dabfb162-0a63-44c0-8c91-ba3073554bc2.png)

The delay between the last bit of the stream and the output going high is 10.6us, against the 100+ us of the ESP32.
The bit width is 1.24us, basically identical to the ESP32's

That is it. Nothing much to add so I think it is my last contribution to this topic. :smiley: . 


## PS: just for completeness, test code
It is the same for both ESP32 and Pico, except for the GPIO pins (0 and 1 for the latter).

```
import time
import machine, neopixel

BR = 5
pattern = [
    (BR, 0, 0),
    (0, BR, 0),
    (0, 0, BR),
    (BR, BR, 0),
    (BR, 0, BR),
    (0, BR, BR),
    (BR, BR, BR),
    (0, 0, 0),
    ]

def blank(np = None):
    if not np:
        led = machine.Pin(13, machine.Pin.OUT)
        np = neopixel.NeoPixel(led, 24, timing = 1)
    np.fill((0,0,0))
    np.write()
    time.sleep_ms(1)

def run(delay=500, neonum = 24):
    led = machine.Pin(13, machine.Pin.OUT)
    sync = machine.Pin(15, machine.Pin.OUT)
    sync(0)
    np = neopixel.NeoPixel(led, neonum, timing = 1)
    led_dly = int(delay/8)
    blank(np)
    for i in range(16):
        for j in range(neonum):
            np[j] = pattern[(i + j) % len(pattern)]
        sync(1)
        np.write()
        sync(0)
        led(1)
        for l in range(8):
            led(1)
            time.sleep_ms(led_dly)
            led(0)
            time.sleep_ms(led_dly)

blank()

```
